java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java

```diff
@@ -33,101 +38,140 @@
-            ArrowType.FloatingPoint fpType = (ArrowType.FloatingPoint) type;
-            switch (fpType.getPrecision()) {
-                case DOUBLE:
-                    return equalEnough((Double)o1, (Double)o2);
-                case SINGLE:
-                    return equalEnough((Float)o1, (Float)o2);
-                case HALF:
-                default:
-                    throw new UnsupportedOperationException("unsupported precision: " + fpType);
-            }
-        } else if (type instanceof ArrowType.Binary) {
-            return Arrays.equals((byte[]) o1, (byte[]) o2);
-        }
+    Comparator<DictionaryEncoding> comp = new Comparator<DictionaryEncoding>() {
+      @Override
+      public int compare(DictionaryEncoding enc1, DictionaryEncoding enc2) {
+        return enc1.getId() < enc2.getId() ? -1 : 1;
+      }
+    };
+
+    // Order of encodings not important
+    Collections.sort(encodings1, comp);
+    Collections.sort(encodings2, comp);
 
-        return Objects.equal(o1, o2);
+    for (int i = 0; i < encodings1.size(); i++) {
+      if (!encodings1.get(i).equals(encodings2.get(i))) {
+        throw new IllegalArgumentException("Different dictionary encodings:\n" + encodings1.get(i) + "\n" + encodings2.get(i));
+      }
+
+      long id = encodings1.get(i).getId();
+      Dictionary dict1 = provider1.lookup(id);
+      Dictionary dict2 = provider2.lookup(id);
+
+      if (dict1 == null || dict2 == null) {
+        throw new IllegalArgumentException("The DictionaryProvider did not contain the required dictionary with id: " + id +"\n" + dict1 + "\n" + dict2);
+      }
+
+      if (!dict1.equals(dict2)) {
+        throw new IllegalArgumentException("Different dictionaries:\n" + dict1 + "\n" + dict2);
+      }
     }
+  }
 
-    static boolean equalEnough(Float f1, Float f2) {
-        if (f1 == null || f2 == null) {
-            return f1 == null && f2 == null;
-        }
-        if (f1.isNaN()) {
-            return f2.isNaN();
-        }
-        if (f1.isInfinite()) {
-            return f2.isInfinite() && Math.signum(f1) == Math.signum(f2);
+  /**
+   * Validate two arrow vectorSchemaRoot are equal.
+   *
+   * @param root1 the 1st shema to compare
+   * @param root2 the 2nd shema to compare
+   * @throws IllegalArgumentException if they are different.
+   */
+  public static void compareVectorSchemaRoot(VectorSchemaRoot root1, VectorSchemaRoot root2) {
+    compareSchemas(root2.getSchema(), root1.getSchema());
+    if (root1.getRowCount() != root2.getRowCount()) {
+      throw new IllegalArgumentException("Different row count:\n" + root1.getRowCount() + "\n" + root2.getRowCount());
+    }
+    List<FieldVector> arrowVectors = root1.getFieldVectors();
+    List<FieldVector> jsonVectors = root2.getFieldVectors();
+    if (arrowVectors.size() != jsonVectors.size()) {
+      throw new IllegalArgumentException("Different column count:\n" + arrowVectors.size() + "\n" + jsonVectors.size());
```

Comment:
add arrowVectors and jsonVectors to the message? This would help with understanding why the sizes are different.